pam_env: use vendor specific pam_env.conf and environment as fallback

schubi2 opened the pull request:
Use the vendor directory (/usr/etc) as fallback for a distribution provided default config if there is no one in /etc.

* Makefile.am: Add libeconf setting.

* pam_env.c: Take care about the fallback configuration in /usr/etc.

* pam_env.8.xml: Add /usr/etc description.

* pam_env.conf.5.xml: Add /usr/etc description.

* tst-pam_env-retval.c: General testcases for pam_env.

* tests: Contains data for testcases.

ldv-alt commented:
Could you make sure it passes CI, please?

thkukuk commented on modules/pam_env/pam_env.8.xml:
```diff
@@ -52,13 +52,26 @@
-    <para>
+    <para condition="with_vendordir">
+      Second a file (<filename>/usr/etc/environment</filename> by default) with simple
+      <emphasis>KEY=VAL</emphasis> pairs on separate lines will be read.
+      An alternate file is <filename>/etc/security/pam_env.conf</filename>
+      or can be specifed with the <emphasis>envfile</emphasis> option.
+      And with the <emphasis>readenv</emphasis> option this can be completely disabled.
+    </para>
```
Please also document /etc/security/pam_env.conf.d for the case we use libeconf, so that people know about all new features.

schubi2 replied:
I have updated the docu.

thkukuk commented on modules/pam_rootok/pam_rootok.c:
```diff
@@ -53,11 +53,10 @@ static int
 PAM_FORMAT((printf, 2, 3))
 log_callback (int type UNUSED, const char *fmt, ...)
 {
-    int audit_fd;
     va_list ap;
 
 #ifdef HAVE_LIBAUDIT
-    audit_fd = audit_open();
+    int audit_fd = audit_open();
 
     if (audit_fd >= 0) {
 	char *buf;
```
Please make this fix an own PR, so that there is no conflict with the other PR containing the same fix.

ldv-alt commented:
Stefan, could you rebase and squash commits, please?

t8m commented:
Please do not add merge commits. Could you please rebase instead?

schubi2 commented:
> Stefan, could you rebase and squash commits, please?

Yes, I have done it...hopefully correctly :-)

schubi2 commented:
ups, sorry

ldv-alt commented on modules/pam_rootok/pam_rootok.c:
```diff
@@ -53,10 +53,11 @@ static int
 PAM_FORMAT((printf, 2, 3))
 log_callback (int type UNUSED, const char *fmt, ...)
 {
+    int audit_fd;
     va_list ap;
 
 #ifdef HAVE_LIBAUDIT
-    int audit_fd = audit_open();
+    audit_fd = audit_open();
 
     if (audit_fd >= 0) {
 	char *buf;
```
No, this reverts #425.

ldv-alt commented on modules/pam_env/pam_env.8.xml:
```diff
@@ -52,13 +52,40 @@
     </para>
-    <para>
+    <para condition="with_vendordir">
+      Rules for (un)setting of variables can be defined in an own conig
+      file. The path of this file can be specified with the
+      <emphasis>conffile</emphasis> option.
+      If this file has not been defined the default rules are taken from the
+      config files <filename>/etc/security/pam_env.conf</filename> and
+      <filename>/etc/security/pam_env.conf.d/*.conf</filename>.
+      If the file <filename>/etc/security/pam_env.conf</filename> is not
+      available the rules are taken from the files
+      <filename>/usr/etc/security/pam_env.conf</filename>,
+      <filename>/usr/etc/security/pam_env.conf.d/*.conf</filename> and
+      <filename>/etc/security/pam_env.conf.d/*.conf</filename> in that order.
+    </para>
+    <para condition="without_vendordir">
       By default rules for (un)setting of variables are taken from the
       config file <filename>/etc/security/pam_env.conf</filename>. An
       alternate file can be specified with the <emphasis>conffile</emphasis>
       option.
     </para>
-    <para>
+    <para condition="with_vendordir">
+      Environment variables can be defined in file with simple <emphasis>KEY=VAL</emphasis>
+      pairs on separate lines. The path of this file can be specified with the
+      <emphasis>envfile</emphasis> option.
+      If this file has not been defined the settings are taken from the
+      files <filename>/etc/security/environment</filename> and
+      <filename>/etc/security/environment.d/*</filename>.
+      If the file <filename>/etc/environment</filename> is not available the
+      settings are taken from the files <filename>/usr/etc/environment</filename>,
+      <filename>/usr/etc/environment.d/*</filename> and
+      <filename>/etc/environment.d/*</filename> in that order.
+      And last but not least with the <emphasis>readenv</emphasis> option this can
+      be completely disabled.
+    </para>
+    <para condition="without_vendordir">
```
This creates an impression that all these new files are supported when linux-pam is configured using `--enable-vendordir`, but the implementation does something completely different: the only place where `VENDORDIR` is going to be used is a new function `_econf_read_file` which is naturally under `USE_ECONF` guard.

In other words, this PR is not about `--enable-vendordir` support in pam_env but rather about `--enable-econf` support in pam_env, and about new features supported only if linux-pam is configured using `--enable-econf`.

I don't see why `--enable-econf` should be a prerequisite for `--enable-vendordir`, it certainly hasn't been discussed.

thkukuk commented:
@ldv-alt @t8m as all changes requests are solved, can we merged this?

ldv-alt commented:
Sorry, I intended to do the next review of this and other PRs, but putin's war in Ukraine disrupted everything.
Please bear with me, I'll have a look.

ldv-alt commented on configure.ac:
```diff
@@ -512,6 +512,11 @@ if test -n "$enable_vendordir"; then
   AC_DEFINE_UNQUOTED([VENDOR_SCONFIGDIR], ["$enable_vendordir/security"],
 		     [Directory for PAM modules distribution provided configuration files])
   STRINGPARAM_VENDORDIR="--stringparam vendordir '$enable_vendordir' --stringparam profile.condition 'with_vendordir'"
+  if test "$WITH_ECONF" = "yes" ; then
+    STRINGPARAM_VENDORDIR="--stringparam vendordir '$enable_vendordir' --stringparam profile.condition 'with_vendordir;with_vendordir_and_with_econf'"
+  else
+    STRINGPARAM_VENDORDIR="--stringparam vendordir '$enable_vendordir' --stringparam profile.condition 'with_vendordir;with_vendordir_and_without_econf"
+  fi
```
The first definition of `STRINGPARAM_VENDORDIR` is redefined inside the added `if` statement.

schubi2 replied:
Yes, fixed.

ldv-alt commented on modules/pam_env/pam_env.conf.5.xml:
```diff
@@ -20,7 +20,15 @@
-    <para>
+    <para condition="with_vendordir">
+      The <filename>/usr/etc/security/pam_env.conf</filename> and
+      <filename>/etc/security/pam_env.conf</filename> file specifies
+      the environment variables to be set, unset or modified by
+      <citerefentry><refentrytitle>pam_env</refentrytitle><manvolnum>8</manvolnum></citerefentry>.
+      When someone logs in, this file is read and the environment
```
While one part of the statement talks about two files, another part talks about a single file, please check the grammar.

ldv-alt commented on modules/pam_env/pam_env.conf.5.xml:
```diff
@@ -61,7 +69,15 @@
 
-    <para>
+    <para condition="with_vendordir">
+      The <filename>/usr/etc/environment</filename> and <filename>/etc/environment</filename> file specifies
+      the environment variables to be set. The file must consist of simple
```
Likewise, please check the grammar.

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -53,7 +62,15 @@ typedef struct var {
 #define ILLEGAL_VAR  103
 
+#ifdef USE_ECONF
+static int _isDirectory(const char *path);
+static char **_econf_read_file(const char *filename, const pam_handle_t *pamh,
+			       const char *delim, const char *name,
+			       const char *suffix, const char *subpath);
+#else
+static char **_read_file(const char*filename, const pam_handle_t *pamh);
```
All identifiers starting with underscore is a reserved namespace.
The reason why we used them was static linking we used to support long time ago.
Nowadays this is just a relic from the past.

That is, you don't _have to_ introduce new identifiers from the reserved namespace.

ldv-alt replied:
In pam functions, `pamh` argument is usually the first one.

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -806,11 +993,8 @@ handle_env (pam_handle_t *pamh, int argc, const char **argv)
-  if(readenv && retval == PAM_SUCCESS) {
+  if(readenv && retval == PAM_SUCCESS)
     retval = _parse_env_file(pamh, ctrl, env_file);
-    if (retval == PAM_IGNORE)
-      retval = PAM_SUCCESS;
```
Why do you remove these two lines?  Is this a bug fix, or is it a change in behaviour?

ldv-alt reviewed (commented):
Stefan, when you are moving a lot of old code around along with new changes, a reviewer has to analyze the whole mix, which is an extra burden that I'd like to avoid. If you really want to do some refactoring, please separate these changes from those that change behaviour.

ldv-alt commented on modules/pam_env/tst-pam_env-retval.c:
```diff
@@ -0,0 +1,273 @@
+
+  setup();
+
+  /* PAM_SUCCESS -> PAM_SUCCESS, PAM_IGNORE -> PAM_PERM_DENIED */
+  ASSERT_NE(NULL, fp = fopen(service_file, "w"));
+  ASSERT_LT(0, fprintf(fp, "#%%PAM-1.0\n"
+		       "auth required %s/.libs/%s.so\n"
+		       "account required %s/.libs/%s.so\n"
+		       "password required %s/.libs/%s.so\n"
+		       "session required %s/.libs/%s.so\n",
```
This is undefined behaviour: `pam_env` returns `PAM_SUCCESS` when environment variables are set.
Please have a look at [ldv/tst-pam_env-retval](https://github.com/linux-pam/linux-pam/commit/b53b9511274d2d98b2055a9b4ab7e635d3c9727a).

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -375,6 +353,215 @@ static int _assemble_line(FILE *f, char *buffer, int buf_len)
+	(void) fclose(conf);
+	return NULL;
+      }
+      lines[i-1] = strdup(buffer);
+    }
+    /* realloc one extra element for the last 0 */
+    lines = realloc(lines, ++i * sizeof(char**));
```
Every memory allocation function can fail any time.

But why reading the whole file into memory in the first place?

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -375,6 +353,215 @@ static int _assemble_line(FILE *f, char *buffer, int buf_len)
+    size_t i = 0;
+    while (_assemble_line(conf, buffer, BUF_SIZE) > 0) {
+      D(("Read line: %s", buffer));
+      lines = realloc(lines, ++i * sizeof(char**));
+      if (lines == NULL) {
+	pam_syslog(pamh, LOG_ERR, "Cannot allocate memory.");
+	(void) fclose(conf);
+	return NULL;
+      }
```
This is a memory leak on error path.

thkukuk commented:
@ldv-alt @t8m If nobody objects I will merge next week.

ldv-alt commented:
I'll have a look this weekend.

ldv-alt reviewed (changes requested):
The main problem with this change is that you are moving big functions (`_parse_config_file` and `_parse_env_file`) and modifying them in the same commit, this makes reviewing changes impossible, please don't do that.

ldv-alt commented on modules/pam_env/tst-pam_env-retval.c:
```diff
@@ -17,11 +17,18 @@
 static const char service_file[] = TEST_NAME ".service";
 static const char missing_file[] = TEST_NAME ".missing";
-static const char my_conf[] = TEST_NAME ".conf";
-static const char my_env[] = TEST_NAME ".env";
+static const char dir[] = TEST_NAME_DIR;
+static const char dir_usr[] = TEST_NAME_DIR "/usr";
+static const char dir_usr_etc[] = TEST_NAME_DIR "/usr/etc";
+static const char dir_usr_etc_security[] = TEST_NAME_DIR "/usr/etc/security";
+static const char my_conf[] = TEST_NAME_DIR ".conf";
+static const char my_env[] = TEST_NAME_DIR ".env";
```
I don't see why `my_conf` and `my_env` are changed here.

ldv-alt commented on modules/pam_env/tst-pam_env-retval.c:
```diff
@@ -41,13 +53,31 @@ setup(void)
+
+	ASSERT_NE(NULL, fp = fopen(usr_conf, "w"));
+	ASSERT_LT(0, fprintf(fp,
+			     "PAGER		DEFAULT=emacs\n"
+			     "MANPAGER	DEFAULT=less\n"));
```
Please use `\t` instead of `	` in C strings.

ldv-alt commented on modules/pam_env/tst-pam_env-retval.c:
```diff
@@ -191,6 +221,36 @@ main(void)
 
+#if defined (USE_ECONF)	&& defined (VENDORDIR)
+
+	/* envfile is a directory. So values will be read from /var/etc and /etc */
```
From /var/etc?
Do we check that both directories are being used?

ldv-alt commented on modules/pam_env/tst-pam_env-retval.c:
```diff
@@ -191,6 +221,36 @@ main(void)
+	const char *env3[] = {"var_etc_test=foo", "var_etc_test2=bar", NULL};
+	check_env(env3);
+
+	/* conffile is a directory. So values will be read from /var/etc and /etc */
```
Likewise.

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -128,160 +157,132 @@ _pam_parse (const pam_handle_t *pamh, int argc, const char **argv,
-		continue;
+    if (filename == NULL || strlen(base_dir) > 0) {
+      /* Read and merge all setting in e.g. /usr/etc and /etc */
+      char *vendor_dir, *sysconf_dir;
```
Initializing `vendor_dir` to NULL would simplify the code below.

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -128,160 +157,132 @@ _pam_parse (const pam_handle_t *pamh, int argc, const char **argv,
-		           "missing key name '%s' in %s', ignoring",
-		           key, file);
-		continue;
+    if (filename == NULL || strlen(base_dir) > 0) {
```
`base_dir[0] != '\0'` would be enough

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -128,160 +157,132 @@ _pam_parse (const pam_handle_t *pamh, int argc, const char **argv,
+    if (filename == NULL || strlen(base_dir) > 0) {
+      /* Read and merge all setting in e.g. /usr/etc and /etc */
+      char *vendor_dir, *sysconf_dir;
+      if (subpath != NULL && strlen(subpath) > 0) {
```
`subpath[0] != '\0'` would be enough

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -128,160 +157,132 @@ _pam_parse (const pam_handle_t *pamh, int argc, const char **argv,
+			      delim, "#");
+      free(vendor_dir);
+      free(sysconf_dir);
+      free(base_dir);
```
Why do you need to allocate memory for `base_dir` in the first place?

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -128,160 +157,132 @@ _pam_parse (const pam_handle_t *pamh, int argc, const char **argv,
+      if (error != ECONF_SUCCESS) {
+	pam_syslog(pamh, LOG_ERR, "Unable to read configuration in different directories: %s",
+		   econf_errString(error));
+	return PAM_ABORT;
```
I'm not sure `PAM_ABORT` is an appropriate return code here.

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -128,160 +157,132 @@ _pam_parse (const pam_handle_t *pamh, int argc, const char **argv,
-	   actually set first, so we don't get a vague error back from
-	   pam_putenv() */
-	for (i = 0; key[i] != '=' && key[i] != '\0'; i++);
+    *lines = malloc((key_number +1)* sizeof(char**));
+    (*lines)[key_number] = 0;
```
Every memory allocation is allowed to fail.

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -128,160 +157,132 @@ _pam_parse (const pam_handle_t *pamh, int argc, const char **argv,
+        if (asprintf(&(*lines)[i],"%s%c%s", keys[i], delim[0], val) < 0) {
+	  pam_syslog(pamh, LOG_ERR, "Cannot allocate memory.");
+	  free_string_array(*lines);
+	  return PAM_BUF_ERR;
```
This leaks both `keys` and `key_file`.

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -375,6 +376,221 @@ static int _assemble_line(FILE *f, char *buffer, int buf_len)
+    }
+
+    size_t i = 0;
+    *lines = malloc((i + 1)* sizeof(char**));
+    (*lines)[i] = 0;
```
Every memory allocation is allowed to fail.

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -375,6 +376,221 @@ static int _assemble_line(FILE *f, char *buffer, int buf_len)
     return used;
 }
 
+static int read_file(const pam_handle_t *pamh, const char*filename, char ***lines)
```
Why do you want to read the whole file into memory?

ldv-alt reviewed (changes requested):
Stefan, I asked you several times to avoid mixing code reordering with functional changes, but, unfortunately, you failed to do that.  Now I did the reordering, please rebase on top of the master branch.

schubi2 commented:
> Stefan, I asked you several times to avoid mixing code reordering with functional changes, but, unfortunately, you failed to do that. Now I did the reordering, please rebase on top of the master branch.

Sorry for the additional work Dmitry. Somehow I have not really gotten your point. Now I see it. I have moved functions in the code and have changed them in one PR. That's really too much. Sorry again, next time I will ask earlier how to proceed. 
I have rebased the code and have not touched the position of the functions. I hope it is a little bit better for reading it. But the code could be still complex due the three code stream:
- as it is
- with vendor handling
- with libeconf (with and without vendor handling)

thkukuk commented:
If there are no more objections I will merge this the next days.

ldv-alt commented:
I'll try to have a look again tomorrow.

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -119,6 +138,154 @@ _pam_parse (const pam_handle_t *pamh, int argc, const char **argv,
+      return PAM_ABORT;
+    }
+
+    *lines = malloc((key_number +1)* sizeof(char**));
+    if (lines == NULL) {
```
If `*lines` has been assigned, then `lines` is definitely non-NULL, you probably meant to check `*lines` instead.

schubi2 replied:
Yes, you are right

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -212,6 +379,52 @@ _assemble_line(FILE *f, char *buffer, int buf_len)
+    }
+
+    size_t i = 0;
+    *lines = malloc((i + 1)* sizeof(char**));
+    if (lines == NULL) {
```
Likewise, if `*lines` has been assigned, then `lines` is definitely non-NULL, you probably meant to check `*lines` instead.

schubi2 replied:
Yes, you are right

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -128,160 +157,132 @@ _pam_parse (const pam_handle_t *pamh, int argc, const char **argv,
+    econf_err error;
+    size_t key_number = 0;
+    char **keys = NULL;
+    char *base_dir = strdup("");
```
I wonder why do you need to have a copy of a string in `base_dir`.
Why can't you just have
```
const char *base_dir = "";
...
base_dir = filename;
```
This way you won't have to deal with memory allocation issues.

schubi2 replied:
Yes, looks much more better now !

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -119,6 +138,138 @@ _pam_parse (const pam_handle_t *pamh, int argc, const char **argv,
+		   keys[i],
+		   econf_errString(error));
+      } else {
+        if (asprintf(&(*lines)[i],"%s%c%s", keys[i], delim[0], val) < 0) {
+	  pam_syslog(pamh, LOG_ERR, "Cannot allocate memory.");
+          econf_free(keys);
+          econf_freeFile(key_file);
+	  free_string_array(*lines);
+	  return PAM_BUF_ERR;
+	}
+	free (val);
```
Note that `val` is leaked here on error path.

schubi2 replied:
yes, good point !

schubi2 commented on modules/pam_env/pam_env.8.xml:
```diff
@@ -52,13 +52,55 @@
+      If this file does not exist <filename>%vendordir%/security/pam_env.conf</filename> is used.
+      An alternate file can be specified with the <emphasis>conffile</emphasis>
+      option, which overrules all other files.
+    </para>
```
> In many places: grammar: path **to** this file
> 
> Please update the whole text from the latest editions you made for other modules.

Hm, the behaviour  differs form the other modules, so I cannot take over from them. But I have tried to improve it.

aplanas commented on modules/pam_env/pam_env.8.xml:
```diff
@@ -52,13 +52,55 @@
     </para>
-    <para>
+    <para condition="with_vendordir_and_with_econf">
+      Rules for (un)setting of variables can be defined in an own conig
```
Typo: conig -> config

schubi2 replied:
fixed. Thank you

ldv-alt commented:
Looks like the only remaining concern is that the new implementation unconditionally loads the whole file (or files) into memory.
I'm not sure it's a good idea, but I'd like to hear other opinions.

thkukuk commented:
> Looks like the only remaining concern is that the new implementation unconditionally loads the whole file (or files) into memory. I'm not sure it's a good idea, but I'd like to hear other opinions.

Which problems do you expect? I would assume we are speaking about some kb here, not even MB or GB...

ldv-alt commented:
While these files are expected to be small, we haven't relied on them to be small before, and with this change we would.
It's usually a good idea to make sure that something we rely upon really meets our expectations.

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -682,19 +886,36 @@ static int
+#ifndef USE_ECONF
+    /* Only one file will be parsed. So, file has to be set. */
+    if (file == NULL) /* No filename has been set via argv. */
+      file = DEFAULT_ETC_ENVFILE;;
```
Two semicolons in a row?

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -625,34 +823,41 @@ static int
+    /* If "file" is not NULL, only this file will be parsed. */
+    retval = econf_read_file(pamh, file, " \t", PAM_ENV, ".conf", "security", &conf_list);
+#else
+    retval = read_file(pamh, file, &conf_list);
```
I think the `#ifdef VENDOR_DEFAULT_CONF_FILE` hunk could be moved right before the `read_file` invocation, this would make `#ifndef USE_ECONF` there unneeded.

schubi2 replied:
Yes, good point

ldv-alt commented on modules/pam_env/pam_env.c:
```diff
@@ -682,19 +886,36 @@ static int
+    char *key, *mark;
+    char **env_list = NULL;
+
+#ifdef VENDOR_DEFAULT_ETC_ENVFILE
+#ifndef USE_ECONF
+    /* Only one file will be parsed. So, file has to be set. */
+    if (file == NULL) /* No filename has been set via argv. */
+      file = DEFAULT_ETC_ENVFILE;;
+    /*
+    * Check whether DEFAULT_ETC_ENVFILE; file is available.
+    * If it does not exist, fall back to VENDOR_DEFAULT_ETC_ENVFILE; file.
+    */
+    struct stat stat_buffer;
+    if (stat(file, &stat_buffer) != 0 && errno == ENOENT) {
+      file = VENDOR_DEFAULT_ETC_ENVFILE;
+    }
+#endif
+#endif
```
In `USE_ECONF` case this means that `file` can be `NULL`, and this is problematic because `file` is used later in this function for error diagnostics.

schubi2 replied:
I am ensuring now that file is not NULL while reading the file via the "old style".

ldv-alt commented:
The problem with this workflow is that when requested changes take long time to implement, the reviewer no longer remembers what those changes were about, and has to do another review almost from scratch.

thkukuk commented on modules/pam_env/pam_env.c:
```diff
@@ -119,6 +138,139 @@ _pam_parse (const pam_handle_t *pamh, int argc, const char **argv,
+      free(vendor_dir);
+      free(sysconf_dir);
+      if (error != ECONF_SUCCESS) {
+	pam_syslog(pamh, LOG_ERR, "Unable to read configuration in different directories: %s",
+		   econf_errString(error));
```
For an user it is not clear which configuration file is not found: /etc/environment or any variant? Or pam_env.conf?
I think we should change this to something like:
`pam_syslog(pamh, LOG_ERR, "Unable to read configuration \"<configuration file>\": %s", econf_errString(error)); `or something like that.

schubi2 replied:
I have used "econf_errLocation" in libeconf for this.

schubi2 commented on modules/pam_env/pam_env.c:
```diff
@@ -119,6 +138,148 @@ _pam_parse (const pam_handle_t *pamh, int argc, const char **argv,
+      free(sysconf_dir);
+      if (error != ECONF_SUCCESS) {
+        if (error == ECONF_NOFILE) {
+          pam_syslog(pamh, LOG_ERR, "No configuration file found");
```
Yes, I have updated it.

schubi2 commented on modules/pam_env/pam_env.c:
```diff
@@ -119,6 +138,148 @@ _pam_parse (const pam_handle_t *pamh, int argc, const char **argv,
+      free(sysconf_dir);
+      if (error != ECONF_SUCCESS) {
+        if (error == ECONF_NOFILE) {
+	  pam_syslog(pamh, LOG_ERR, "Configuration file not found: %s.%s", name, suffix);
```
Yes, I have reverted it.

ldv-alt commented on configure.ac:
```diff
@@ -523,6 +523,7 @@ if test "$WITH_ECONF" = "yes" ; then
   PKG_CHECK_MODULES([ECONF], [libeconf], [],
   [AC_CHECK_LIB([econf],[econf_readDirs],[ECONF_LIBS="-leconf"],[ECONF_LIBS=""])])
   if test -n "$ECONF_LIBS" ; then
+    AC_CHECK_LIB([econf],[econf_errLocation], [], [AC_MSG_ERROR([Please update libeconf])])
```
This configure snippet requires that `libeconf` either doesn't exist or is fresh enough, which is somewhat unusual.
I suppose it would be surprising for many to see the build fail by default in case` libeconf` is installed but not fresh enough.

schubi2 replied:
I have changed it to check the version of libeconf only.

ldv-alt commented on configure.ac:
```diff
@@ -520,8 +520,7 @@ AC_ARG_ENABLE([econf],
 if test "$WITH_ECONF" = "yes" ; then
-  PKG_CHECK_MODULES([ECONF], [libeconf], [],
-  [AC_CHECK_LIB([econf],[econf_readDirs],[ECONF_LIBS="-leconf"],[ECONF_LIBS=""])])
+  PKG_CHECK_MODULES([ECONF], [libeconf >= 0.5.0], [ECONF_LIBS="-leconf"], [ECONF_LIBS=""] )
```
I wonder wouldn't a simpler variant be enough, e.g.
```
PKG_CHECK_MODULES([ECONF], [libeconf >= 0.5.0])
```

schubi2 replied:
Yes, that should be done in a more simple way. Fixed.
There is another PR which checks this too. I will fix it too.

schubi2 replied:
Hm, the "problem" is the buid PAM environment. It does not contain libeconf but configure will be called without the option --disable-econf. So it is is so "complicate" because libeconf will be disabled for the build if it is not there. So I have reverted it in order to get a running build again:
+++ b/configure.ac
@@ -520,9 +520,10 @@ AC_ARG_ENABLE([econf],
   AS_HELP_STRING([--disable-econf], [do not use libeconf]),
   [WITH_ECONF=$enableval], WITH_ECONF=yes)
 if test "$WITH_ECONF" = "yes" ; then
-  PKG_CHECK_MODULES([ECONF], [libeconf >= 0.5.0])
-  ECONF_LIBS="-leconf"
-  ECONF_CFLAGS="-DUSE_ECONF=1 $ECONF_CFLAGS"
+  PKG_CHECK_MODULES([ECONF], [libeconf >= 0.5.0], [ECONF_LIBS="-leconf"],[ECONF_LIBS=""])
+  if test -n "$ECONF_LIBS" ; then
+    ECONF_CFLAGS="-DUSE_ECONF=1 $ECONF_CFLAGS"
+  fi
 fi

Either we put libeconf in the PAM build environment, or we disabling libeconf via --disable-econf in the configure calls or we disable libeconf (almost) silently if it is not there (current state in the PR again).
So, what is would be prefered ?

thkukuk replied:
> Sorry, I don't understand what was the problem with `PKG_CHECK_MODULES([ECONF], [libeconf >= 0.5.0])`.

With your suggestion, configure will fail if libeconf is not installed and you can only build Linux-PAM if you use --disable-econf.
Because this statement makes libeconf >= 0.5.0 a hard requirement.
With Stefans solution, the test will not fail, but libeconf support get's disabled if libeconf is not installed.

ldv-alt replied:
I see.  I suppose this would work:
```
ECONF_CFLAGS=
ECONF_LIBS=
AC_ARG_ENABLE([econf],
  AS_HELP_STRING([--disable-econf], [do not use libeconf]),
  [WITH_ECONF=$enableval], [WITH_ECONF=yes])
if test "$WITH_ECONF" = "yes"; then
  PKG_CHECK_MODULES([ECONF], [libeconf >= 0.5.0], [], [])
  if test -n "$ECONF_LIBS"; then
    ECONF_CFLAGS="-DUSE_ECONF=1 $ECONF_CFLAGS"
  fi
fi
AC_SUBST([ECONF_CFLAGS])
AC_SUBST([ECONF_LIBS])
```

By the way, it should have been called --with-econf/--without-econf rather than --enable-econf/--disable-econf.

ldv-alt replied:
Or even a shorter form:
```
ECONF_CFLAGS=
ECONF_LIBS=
AC_ARG_ENABLE([econf],
  AS_HELP_STRING([--disable-econf], [do not use libeconf]),
  [WITH_ECONF=$enableval], [WITH_ECONF=yes])
if test "$WITH_ECONF" = "yes"; then
  PKG_CHECK_MODULES([ECONF], [libeconf >= 0.5.0], [ECONF_CFLAGS="-DUSE_ECONF=1 $ECONF_CFLAGS"], [:])
fi
AC_SUBST([ECONF_CFLAGS])
AC_SUBST([ECONF_LIBS])
```

schubi2 replied:
Thank you !!!!

ldv-alt commented:
Stefan, in the commit message you mix phrases in the present tense (e.g. `Add libeconf setting`) and in the past tense (e.g. `Added tests for libeconf`), please follow the same style in all sentences, preferably the present tense.

lnussel commented:
Is it really necessary to add `environment.d` to pam when there's `pam_env.conf.d`? The former is already used by systemd, see `environment.d(5)`

ldv-alt commented:
> Is it really necessary to add `environment.d` to pam when there's `pam_env.conf.d`? The former is already used by systemd, see `environment.d(5)`

`/etc/security/pam_env.conf` and `/etc/environment` are not to be confused, they have different format, see pam_env(8).
